### ISSUE TITLE
Support for Charts

### ISSUE THREAD

**Issue by sorasful** [huntabyte/shadcn-svelte#1175]
### Describe the feature

Just recently, Shacdn official has launched Charts : https://ui.shadcn.com/docs/components/chart 

I was thinking we can keep this issue open to track when this'll be implemented.  I'm not sure about the way the components are translated from Shacdn official to Svelte. Let me know if I can help.

**Comment by huntabyte** [comment]
I'm in contact with @techniq about how we can integrate `layerchart` in a composable way similar to the original with recharts.

It appears this issue will get us most of the way there:
https://github.com/techniq/layerchart/issues/223

**Comment by techniq** [comment]
![tom-hanks-hello-gif](https://github.com/huntabyte/shadcn-svelte/assets/177476/a8dd4848-93ce-4205-8723-3a083db4cdf5)

**Comment by 0x090909** [comment]
UP UP UP Lets do it

**Comment by techniq** [comment]
Btw, using LayerChart with shadcn-svelte is already possible.  See the integration [issue](https://github.com/techniq/layerchart/issues/160) (which will make it to the docs soon) as well as a simple example [project](https://github.com/techniq/layerchart-shadcn-svelte).  Mostly the integration is just mapping tailwind design tokens/colors (`surface-*` and `primary`) but everything can be overwritten, the tokens just streamline conventions.  

[Simplified Chart components](https://github.com/techniq/layerchart/issues/223) is to simplify common use cases like AreaChart, BarChart, etc,  but you can always leverage the full components and chart types within LayerChart (start with `Chart`, etc).  These simplified chart components will also simplify non-shadcn svelte usage as well especially for data exploration, so it's a win-win.

Not sure when I'll get started on the effort, but I have a good idea of the approach and shouldn't take long, and we can continue to iterate.

**Comment by 0x090909** [comment]
@techniq @huntabyte we want the templates that shadcn have on their page, they make the whole process 100x easier

**Comment by techniq** [comment]
@0x090909 speaking for @huntabyte, but that docs integration will be coming. Mostly exposing the same examples in LayerChart but with shadcn-svelte copy/paste.

**Comment by ddtamn** [comment]
Lets go, i will integrate with my app🙌

**Comment by masayahingpuno** [comment]
LFG!!!

**Comment by julien-blanchon** [comment]
I just want to give love ❤️

**Comment by itsmikesharescode** [comment]
any update with this ??

**Comment by 0x090909** [comment]
updates?

**Comment by 0x090909** [comment]
is there any possibility these charts will be realease in the near future? or we should use another chart library ? 

let us know

**Comment by 0x090909** [comment]
updates?=

**Comment by techniq** [comment]
Update: Sorry work/life has been rather busy (and not slowing down).  I also just returned from a family vacation today, but I hope to resume work on the [Simplified Charts PR](https://github.com/techniq/layerchart/pull/224) this week 🤞.

While the end goal is to have copy/paste support for common chart types (Bar, Line, Area, etc) on the shadcn-svelte [docs](https://www.shadcn-svelte.com/), 80% of this is already available with slightly longer form `<Chart>` usage and a few extra imports so if you don't want to be blocked now, I encourage you to start there in the interim.   It's as simple as:

- Install a few dependencies: `npm install layerchart d3-scale d3-array`
  - Even after simplified charts is complete, `layerchart` will always be a required dependency (similar to recharts with shadcn/ui).  If you stick to these simplified charts, you won't need `d3-scale` or `d3-array` added directly, but you might run across a chart type that doesn't have a simplified version, or you want more control, so it wouldn't hurt to have both of these `d3` dependencies installed going forward.
- Update your `tailwind.config.ts` with `surface-*` colors used as default colors for some LayerChart components (LayerChart also uses `primary` which is already available in shadcn-svelte).  See the example [project](https://github.com/techniq/layerchart-shadcn-svelte)  as a reference
```ts
export default {
  content: [
    "./src/**/*.{html,js,svelte,ts}",
    "./node_modules/layerchart/**/*.{svelte,js}"
  ],
  theme: {
    extend: {
      colors: {
        surface: {
          content: "hsl(var(--card-foreground) / <alpha-value>)",
          100: "hsl(var(--background) / <alpha-value>)",
          200: "hsl(var(---muted) / <alpha-value>)",
          // not sure what color maps here (should be darker than 200).  Could add a new color to `app.css`
          300: "hsl(var(--background) / <alpha-value>)"
        }
      }
    },
  },
};
```
- Find the chart type you would like to use, for example [Area Chart](https://www.layerchart.com/docs/examples/Area), expand the code and click the copy button in the top right.
![image](https://github.com/user-attachments/assets/02511b2d-194a-44a6-90be-464caf178e5d)
- Add the missing imports for the LayerChart components (ex. `import { Chart, Svg, Axis, .... } from 'layerchart'`) and d3 usage (ex. `import { scaleTime } from 'd3-scale`).  A lot of the examples also use Svelte UX's [format](https://svelte-ux.techniq.dev/docs/utils/format) for simplified number and date formatting with Intl support so you'll need to replace this with your own formatting solution.  I plan to extract this functionality from Svelte UX to make it easier to use with LayerChart-only usage in the future.
- Profit!

I don't have a deadline for when the simplified charts PR and subsequent shadcn-svelte docs integration will be complete but it's at the top of my priority queue ATM.  I work on open source mostly in the evenings as work and family schedule allows, but hopefully it won't be much longer.  I also suggest monitoring the [Simplified Charts PR](https://github.com/techniq/layerchart/pull/224) for progress.

Have a great day!

**Comment by simsjack30** [comment]
@techniq you're an absolute machine. Thanks for working on this and keeping us updated

**Comment by RickRyan26** [comment]
So awesome! Thank you all

**Comment by 0x090909** [comment]
Updates?

**Comment by 0x090909** [comment]
these charts are life and death decision

**Comment by sorasful** [comment]
@0x090909 just read 3 comments before yours.

**Comment by 0x090909** [comment]
any updates?

**Comment by techniq** [comment]
Quick update:

I've been very busy cranking away on the new Simplified Charts, refined Tooltips, extracting a new suite of packages (`@layerstack/*`), and a few other goodies...

![image](https://github.com/user-attachments/assets/50bce784-5d27-4ccf-8f5c-2e52dae41b35)

So far I've created simplified charts for `BarChart`, `LineChart`, `AreaChart`, and `ScatterChart`, and you can see the progress in the [PR](https://github.com/techniq/layerchart/pull/224) and [preview docs](https://techniq-simplified-charts.layerchart.pages.dev/docs/components/AreaChart)

![image](https://github.com/user-attachments/assets/78820dfc-791f-4d08-ae95-02c35c0b823c)

There are more use cases for these simplified components I want to think through and support, update more of the existing examples to these components (but trying to find the right balance as well), and add more of these chart types (PieChart, stack chart, radial charts, etc).  See some of my thoughts in the original [issue](https://github.com/techniq/layerchart/issues/223).   The goal of the simplified charts is "streamlined opinionated base chart components with full extension capability" so anything you could have created with `<Chart>` can be accomplished with these components, with less props and imports.  They basically providing you the 80% out of the box, but can still get to 100% using the same props and components you would use with Chart (in some ways you can them of them as Chart templates). 

Some of the stack/group examples might have to wait until more [Chart improvements](https://github.com/techniq/layerchart/issues/183) are integrated (see also LayerCake's derivedScales [PR](https://github.com/mhkeller/layercake/pull/210)).  You can 100% create stacked, grouped, stacked and grouped, multi series, diverging, etc charts now (see all the existing examples), but I want to simplify both the implementation and the usage.

I'm contemplating updating LayerChart's `Chart` component to provide the full functionality of LayerCake's `LayerCake` chart component for a more integration opportunity and simplifications (I've worked around this so far with dedicated components and contexts like [GeoContext](https://github.com/techniq/layerchart/blob/main/packages/layerchart/src/lib/components/GeoContext.svelte), [TooltipContext](https://github.com/techniq/layerchart/blob/main/packages/layerchart/src/lib/components/TooltipContext.svelte), and [TransformContext](https://github.com/techniq/layerchart/blob/main/packages/layerchart/src/lib/components/TransformContext.svelte), and even replaced the [rendering layers](https://github.com/techniq/layerchart/tree/main/packages/layerchart/src/lib/components/layout) (Svg, Canvas, Html) with LayerChart versions to provide this level of integration I couldn't obtain otherwise. 

My work has been pretty crazy lately, and it's also soccer season so I'm always running one of my 3 kids to practices or games this time of year, but anytime I'm near my laptop I'm usually working on this (or thinking about it while exercising) 😁

In short, I'm very actively working on this, and good things will come to those who wait, but don't forget you [don't need to wait](https://github.com/huntabyte/shadcn-svelte/issues/1175#issuecomment-2254800243) on these, and migration will basically be changing `<Chart>` to `<SomeChart>` and removing a bunch of props/components/imports.  I want to both have a strong initial offering once merged, but also plan subsequent iterations based on feedback (adding more simplified charts, etc).

Thanks for your patience 🫶

**Comment by itsmikesharescode** [comment]
nice!!!! layercharts!

**Comment by tkuchh** [comment]
@techniq  thanks for the update! Take your time, spend time with your kids. The community really appreciates all the work, we are happy to wait since it'll definitely be worth it in the end.

**Comment by techniq** [comment]
Tomorrow...

![tommy-wiseau-psa-gif-by-the-room](https://github.com/user-attachments/assets/b89dc553-db14-429a-9c19-3860b13a1dd8)

**Comment by techniq** [comment]
https://github.com/user-attachments/assets/d3f9b88b-82ea-4edb-a081-bece6e57348f

We're live!!!

I wrote up an overview of the masstive [0.50.0](https://github.com/techniq/layerchart/releases/tag/layerchart%400.50.0) release on [Twitter/X](https://x.com/techniq35/status/1838199849608462414) and would appreciate any social love (likes, reposts, etc) to increase reach.

I also [submitted](https://x.com/techniq35/status/1835142292631335221) a talk proposal "What's new with LayerChart" for the fall [Svelte Summit](https://www.sveltesummit.com/2024/fall) conference, and hopefully it gets accepted.

**Comment by cjlryan** [comment]
@techniq properly incredible stuff! Definitely implementing some of this right away! 🙌

**Comment by burggraf** [comment]
@huntabyte is this going to be rolled up into the shadcn-svelte chart component?

**Comment by jdgamble555** [comment]
One of the problems I see with adding packages like this is that they use stores instead of runes. These packages get almost too complex to rewrite with `runes` (although we see @huntabyte did it with svelte-shadcn!) While it can be compatible with Svelte 5, you will be giving up the speed and flexibility of runes. Svelte will eventually drop stores and these libraries (like superforms) will be left behind if they don't get rid of them.

It seems that svelte shad-cn is striving to be rune-only, so IMO spending time adding an outdated package (until it uses runes) would be counter productive.  I personally would like all svelte-shadcn packages to be compatible with rune mode:
```ts
export default {
  compilerOptions: {
    runes: true
  }
};
```

I'm wondering why the charts was removed from @next?

https://github.com/huntabyte/shadcn-svelte/pull/1341

J

**Comment by burggraf** [comment]
I'd prefer, personally, just a port of shadcn chart to shadcn-svelte.

**Comment by techniq** [comment]
@jdgamble555 Don't worry, LayerChart (and Svelte UX) will be migrated to Svelte 5 (runes, snippets) *[soon](https://github.com/techniq/layerchart/issues/159)*...

@burggraf shadcn/ui [charts](https://ui.shadcn.com/charts) use [Recharts](https://recharts.org/en-US/) to provide the chart component foundation (which is React only), and is what LayerChart is providing for Svelte.  This is similar in nature to  shadcn/ui using [radix-ui](https://www.radix-ui.com/) and shadcn-svelte using [bits-ui](https://github.com/huntabyte/bits-ui) (albeit bits-ui is also written by @huntabyte because he's awesome like that).  LayerChart also provides a broader range of visualization components compared to Recharts (cartesian/polar only), such as hierarchy, graph, force, and geo.

I've been supporting @huntabyte and @shyakadavis as they work on the shadcn-svelte integration / docs to make sure LayerChart has all the needed capability, but don't have the bandwidth to do much of this work myself (transitioning jobs this/next week and going hard on Svelte 5 migration soon).

We'll get this one over the finish line, but remember there is already [integration](https://github.com/huntabyte/shadcn-svelte/issues/1175#issuecomment-2254800243) with shadcn-svelte theming with LayerChart that many are already using.

**Comment by itsmikesharescode** [comment]
any update

**Comment by JustKira** [comment]
any updates

**Comment by itsmikesharescode** [comment]
no official guide yet?

**Comment by meoawww** [comment]
Hey @huntabyte any update on charts, when will it be available?

**Comment by 0x090909** [comment]
They left the chat long ago

Apr 28, 2025 20:02:45 M4JOR ***@***.***>:

**Comment by techniq** [comment]
Progressing...

- https://huntabyte-next-tailwind-4-ch.shadcn-svelte.pages.dev/charts
  - https://github.com/huntabyte/shadcn-svelte/pull/1787
- https://github.com/techniq/layerchart/pull/458

**Comment by techniq** [comment]
![Image](https://github.com/user-attachments/assets/49840085-e7b2-414b-8fbb-4af119a557c6)

![Image](https://github.com/user-attachments/assets/dee150e9-8bab-4b98-b16a-30aaaa9e0cbc)
